bittide-instances/src/Bittide/Instances/Hitl/LinkConfiguration.hs

```diff
@@ -0,0 +1,222 @@
+      , "CSB"       ::: Signal Basic125 Bool
+      )
+  )
+linkConfigurationTest refClkDiff sysClkDiff syncIn rxns rxps miso =
+  (txns, txps, syncOut, spiDone, spiOut)
+ where
+  refClk = ibufds_gte3 refClkDiff :: Clock Ext200
+  (sysClk, sysRst) = clockWizardDifferential sysClkDiff noReset
+
+  -- the test starts with a valid configuration coming in
+  startTest = isJust <$> testConfig
+
+  -- use some simple SYNC_IN / SYNC_OUT synchronization to
+  -- synchronously start the test
+  syncInRst =
+      resetGlitchFilter (SNat @1024) sysClk
+    $ unsafeFromActiveLow
+    $ xpmCdcSingle sysClk sysClk syncIn
+  syncOut = startTest
+
+  testRst =   sysRst
+    `orReset` syncInRst
+    `orReset` unsafeFromActiveLow startTest
+    `orReset` unsafeFromActiveHigh skip
+
+  -- derive the test settings from the passed configuration
+  skip = maybe False (not . fpgaEnabled) <$> testConfig
+  mask = maybe 0 disabledLinkMask <$> testConfig
+
+  (txns, txps, linkUps, _stats, spiDone, spiOut) =
+    transceiversStartAndObserve refClk sysClk testRst rxns rxps miso
+
+  -- Consider test done if enabled links have been up consistently for
+  -- 50 seconds. This is just below the test timeout of 60 seconds, so
+  -- transceivers have ~10 seconds to come online reliably. This
+  -- should be plenty.
+  endSuccess = trueFor (SNat @(Seconds 50)) sysClk testRst linkMaskMatch
+
+  linkMaskMatch = and . fmap bitToBool . bv2v <$> (xor <$> mask <*> linkUps)
+
+  testConfig :: Signal Basic125 (Maybe TestConfig)
+  testConfig =
+    let
+      (start, dat) = unbundle $ setName @"vioHitlt" $ vioProbe
+        ("probe_test_done" :> "probe_test_success" :> "linksUps" :>  Nil)
+        ("probe_test_start" :> "probe_test_data" :> Nil)
+        (False, disabled)
+        sysClk
+        -- done
+        (skip .||. endSuccess)
+        -- success
+        (skip .||. linkMaskMatch)
+        -- other
+        linkUps
+    in
+      mux start (Just <$> dat) (pure Nothing)
+
+-- XXX: We use an explicit top entity annotation here, as 'makeTopEntity'
+--      generates warnings in combination with 'Vec'.
+{-# ANN linkConfigurationTest Synthesize
+  { t_name = "linkConfigurationTest"
+  , t_inputs =
+    [ (PortProduct "SMA_MGT_REFCLK_C") [PortName "p", PortName "n"]
+    , (PortProduct "SYSCLK_300") [PortName "p", PortName "n"]
+    , PortName "SYNC_IN"
+    , PortName "GTH_RX_NS"
+    , PortName "GTH_RX_PS"
+    , PortName "MISO"
+    ]
+  , t_output =
+    (PortProduct "")
+      [ PortName "GTH_TX_NS"
+      , PortName "GTH_TX_PS"
+      , PortName "SYNC_OUT"
+      , PortName "spiDone"
+      , (PortProduct "") [PortName "SCLK", PortName "MOSI", PortName "CSB"]
+      ]
+  } #-}
+
+tests :: HitlTests TestConfig
+tests = Map.fromList
+  [ ( "disable " <> fromString (show i) <> " [" <> fromString fpgaId <> "]"
+    , ( toList $ imap (testData i) $ linkMasks @FpgaCount g
+      , NoPostProcData
+      )
+    )
+  | (i, (fpgaId, _)) <- toList $ zip indicesI fpgaSetup
```

Comment:
```suggestion
  | (i, (fpgaId, _)) <- toList $ zip (iterateI succ 0) fpgaSetup
```

That way you can remove the `fromEnum` / ``` i` ```, which makes this a little bit clean IMO. (But I'm totally not hung up on this.)